recipes/graalpy/meta.yaml

```diff
@@ -0,0 +1,199 @@
+    sha256: 9b714747fbff0b9f477a102a22e6558f697322d8fa89b1ac9956938d3afb3b87
+    folder: graal
+    patches:
+      - patches/sulong-disable-tests.patch
+      - patches/sulong-relative-rpath.patch
+  - url: https://github.com/oracle/graalpython/archive/refs/tags/vm-ce-{{ version }}.zip
+    sha256: 72ad9a6e8d8283425768c2838d9f360a42eb1dc6096d0efe44464f4fd5c0982c
+    folder: graalpython
+    patches:
+      - patches/graalpy-honour-sysroot.patch
+      - patches/graalpy-reduce-image-build-xmx.patch
+      - patches/graalpy-fix-lzma.patch  # [osx]
+
+build:
+  number: {{ build_num }}
+  skip: true  # [win or osx]
+
+outputs:
+  - name: graalpy
+    version: {{ version }}
+    script: build_graalpy.sh
+    build:
+      skip_compile_pyc:
+        - "*.py"
+      number: {{ build_num }}
+      string: {{ build_num }}_graalpy{{ python_maj_min.replace(".", "") }}_{{ graalpy_distribution }}
+      track_features:  # [graalpy_distribution == "graalvm"]
+        - openjdk  # [graalpy_distribution == "graalvm"]
+      script_env:  # [graalpy_distribution == "standalone"]
+        - GRAALPY_STANDALONE_BUILD=true  # [graalpy_distribution == "standalone"]
+      run_exports:
+        weak:
+          - graalpy >={{ version }}
+        noarch:
+          - {{ pin_subpackage("python", max_pin="x.x", min_pin="x.x") }}
+    requirements:
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        - pypy3.9 >=7.3.9
+        - make
+        - patch
+        - zlib
+        - git
+        - curl
+        - cmake
+        - patchelf  # [linux]
+      host:
+        - zlib
+      run:
+        - libzlib
+        - patch
+        - curl
+      run_constrained:
+        - __osx >={{ MACOS_DEPLOYMENT_TARGET|default("10.10") }}  # [osx]
+        - {{ pin_subpackage("python", exact=True) }}
+        - {{ pin_subpackage("python_abi", exact=True) }}
+    test:
+      commands:
+        - python3 --help
+        - python3 -c "import platform; print(platform._sys_version())"
+        - graalpy -m venv --help
+        - graalpy -m venv testvenv
+        - testvenv/bin/graalpy -c "import _struct; print(_struct.pack('I', 15663050))"
+
+  - name: python_abi
+    version: {{ python_maj_min }}
+    build:
+      number: {{ build_num }}
+      string: {{ build_num }}_graalpy{{ graalpy_abi }}
+      track_features:
+        - graalpy
+    requirements:
+      run:
+        - {{ pin_subpackage("graalpy", max_pin="x.x.x", min_pin="x.x.x") }}
+      run_constrained:
+        - pypy <0a0
+        - {{ pin_subpackage("python", exact=True) }}
+    test:
+      commands:
+        - graalpy --version
+        - python --version
+
+  - name: python
+    version: {{ gp_python_version }}
+    build:
+      number: {{ build_num }}
+      string: {{ build_num }}_{{ graalpy_abi }}_graalpy
+      track_features:
+        - graalpy
+      run_exports:
+        weak:
+          - {{ pin_subpackage("graalpy", max_pin="x.x", min_pin="x.x") }}
```

Comment:
You need a `python_abi` pin here like pypy.

Reply:
I tried that at some point and then the build failed to resolve complaining about over constraining, but I've changed some other bits since then so I'll try again and see what the CI says.

Reply:
Ok, it works like this.
```yaml
      run_exports:
        weak:
          - {{ pin_subpackage("graalpy", max_pin="x.x", min_pin="x.x") }}
          - {{ pin_subpackage("python_abi", exact=True) }}
```

Reply:
I now used this:
````yaml
      run_exports:
        weak:
          - {{ pin_subpackage("graalpy", max_pin="x.x", min_pin="x.x") }}
          - {{ pin_subpackage("python_abi", max_pin="x.x", min_pin="x.x") }}
```

My reasoning is the following, please let me know if I am still misunderstanding something: The export for graalpy will imply the `run_constrained` spec `{{ pin_subpackage("python_abi", exact=True) }}` from there. So whichever graalpy version is chosen, it will constrain the python_abi to be the exact one that was built with it. Thus, even though we only have a version pin for the `python_abi` in the weak exports for `python` (which by itself may allow python_abi builds pulling in cpython), the `graalpy` weak export ensures that at runtime, users get a compatible python_abi provided via graalpy.

Reply:
I really don't like this complexity. How about you remove `python_abi` from this recipe and we build it in https://github.com/conda-forge/python_abi-feedstock ?

Reply:
I would like to. The problem is what I described [above](https://github.com/conda-forge/staged-recipes/pull/20882#issuecomment-1304795244). As far as I can tell, we cannot pass the CI with this if I already have the `python` output in there, because that needs to export the python_abi with the graalpy_abi in the build string, which is not satisfiable unless we have it first in `python_abi`, but we cannot add it to `python_abi`, yet, because there it needs to depend on `graalpy` which is here. So the only alternative I see is I take all the outputs out of this recipe for now, merge it so we get `graalpy-feedstock`, then open a PR to add the graalpy abi to `python_abi-feedstock`, and then update the graalpy-feedstock to have the `python` output again. What I did here was my attempt at getting this into one PR instead of 3.

Reply:
I pushed the variant where I removed python_abi and commented all references to it out (because they cannot be resolved). If you think this is better, I'm happy with this approach, too :) Then once there's a feedstock for graalpy, I will send the changes to python_abi-feedstock in a PR to that repo and then uncomment the constraints in the graalpy-feedstock and bump the build number there.

Reply:
Since the python_abi PR is now merged and I can thus add the below to constrain to the proper `python_abi` version `python_abi 3.8.* *_graalpy_38_native`

```yaml
run_constrained:
  - python_abi {{ python_maj_min }}.* *_graalpy_{{ python_maj_min.replace('.', '') }}_{{ graalpy_abi }}
```